### PR TITLE
Fix specified destination issuer in pathfinding (RIPD-812)

### PR DESCRIPTION
**Pull request by JoelKatz**
* Compute the effective recipient.
* Make sure the effective recipient exists.
* Prohibit paths to the recipient, if not the effective recipient.
* Treat paths to the effective recipient as complete.
* Don't find looped paths.
* Use the effective recipient for getPathsOut weight.

### CODE REVIEW

File: src/ripple/app/paths/Pathfinder.cpp [hunk 1]
```diff
@@ -175,6 +175,8 @@ Pathfinder::Pathfinder (
     STAmount const& saDstAmount)
     :   mSrcAccount (uSrcAccount),
         mDstAccount (uDstAccount),
+        mEffectiveDst ((saDstAmount.getIssuer() == xrpAccount ()) ?
```
Comment (rec) [inline src/ripple/app/paths/Pathfinder.cpp]: There's a helper function `isXRP()` that should work fine here.

File: src/ripple/app/paths/Pathfinder.cpp [hunk 2]
```diff
@@ -193,6 +195,8 @@ Pathfinder::Pathfinder (
...
```
Comment (rec) [inline src/ripple/app/paths/Pathfinder.cpp]: ditto.

File: src/ripple/app/paths/Pathfinder.h [hunk 3]
```diff
@@ -176,6 +176,7 @@ class Pathfinder
 
     Account mSrcAccount;
     Account mDstAccount;
+    Account mEffectiveDst;
```
Comment (rec) [inline src/ripple/app/paths/Pathfinder.h]: This needs documentation!

What's the difference between the "effective destination" and "destination?

What about `mEffectiveDstAccount` to make it more similar to the other variables?

File: src/ripple/app/paths/Pathfinder.cpp [hunk 4]
```diff
@@ -877,6 +900,10 @@ void Pathfinder::addLink (
     auto const& uEndAccount = pathEnd.getAccountID ();
     bool const bOnXRP = uEndCurrency.isZero ();
 
+    // Does pathfinding really need to get this to
```
Comment (rec) [inline src/ripple/app/paths/Pathfinder.cpp]: This comment is a bit hard to parse (English-wise).  If you had a long comment for `mEffectiveDstAccount` you probably wouldn't need this comment at all...

**Comment by JoelKatz** [comment]
We should do path filtering even if we find just one path to ensure any looped/invalid paths are filtered out.

**Comment by rec** [comment]
@vinniefalco We need another reviewer.

**Comment by nbougalis** [comment]
@rec I'll adopt this. Review will be done by tomorrow morning.

**Comment by yana-novikova** [comment]
Can we remove Tx Change label? It does not affect transaction processing according to David

**Comment by JoelKatz** [comment]
This only sort of needs the API Change tag. The consensus is that it was always supposed to work this way.

**Comment by rec** [comment]
I'm removing the API Change.  

We should reserve it for _breaking_ changes or disruptive changes.  There's no advantage in delaying bug fixes this way...!

**Comment by vinniefalco** [comment]
@rec "API Change" also indicates that partner messaging or documentation updates may be required.

File: src/ripple/app/paths/Pathfinder.h [hunk 5]
```diff
@@ -176,6 +176,7 @@ class Pathfinder
 
     Account mSrcAccount;
     Account mDstAccount;
+    Account mEffectiveDst; // The account the paths need to end at
```
Comment (rec) [inline src/ripple/app/paths/Pathfinder.h]: As I said, I think you should call this `mEffectiveDstAccount` so it's just like `mSrcAccount` and `mDstAccount` (and so you know it's an account when you see it). 

If you just add a little text at the start of this block of variables explaining what the difference is between the "destination account" and the "*effective* destination account" then I'll approve this change for you!

File: src/ripple/app/paths/Pathfinder.cpp (hunk 4 above)
Comment (rec) [inline src/ripple/app/paths/Pathfinder.cpp]: (And then you can just remove the comment here entirely.  :-))

**Comment by rec** [comment]
Ping?

**Comment by rec** [comment]
Ping?

**Comment by rec** [comment]
Ping?

**Comment by nbougalis** [comment]
Looking now. I had reviewed it in past but missed out on commenting and I don't remember if I had any issues.

**Comment by nbougalis** [comment]
:+1:

**Comment by JoelKatz** [comment]
Rebased
